index/index.go

```diff
@@ -0,0 +1,393 @@
+	// Identifier with the largest value of the counter.
+	lastIdentifier, err := i.lastIdentifier(keyword)
+	if err != nil {
+		return err
+	}
+
+	counter := uint64(0)
+
+	// If the current last sealed Identifier is not an empty sealed Identifier, i.e. the keyword
+	// has already been added to the secure index, update the counter.
+	if lastIdentifier.Identifier != "" {
+		counter = lastIdentifier.NextCounter
+	}
+
+	// Compute new labelUUID and plaintext Identifier, seal it, and send it to the IO Provider.
+	labelUUID, err := computeLabelUUID(counter, tagger)
+	if err != nil {
+		return err
+	}
+
+	plainID := data.Identifier{Identifier: identifier, NextCounter: counter + 1}
+
+	sealedID, err := plainID.Seal(labelUUID, cryptor)
+	if err != nil {
+		return err
+	}
+	if err = i.putSealedIdentifier(labelUUID, &sealedID); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Given a keyword, Search returns all decrypted Identifiers that the keyword is contained in.
+func (i *SecureIndex) Search(token, keyword string) ([]string, error) {
+	identity, err := i.idProvider.GetIdentity(token)
+	if err != nil {
+		return nil, ErrNotAuthenticated
+	}
+	if !identity.Scopes.Contains(id.ScopeIndex) {
+		return nil, ErrNotAuthorized
+	}
+
+	tagger, cryptor, err := i.getTaggerAndCryptor(keyword)
+	if err != nil {
+		return nil, err
+	}
+
+	decryptedIDs := []string{}
+
+	// Starting with counter = 0, check if the corresponding keyword label exists in the secure index.
+	// As long as the keyword label exists, decrypt the corresponding Identifier, append it to decryptedIDs,
+	// and update counter to the next counter.
+	counter := uint64(0)
+
+	for {
+		labelUUID, err := computeLabelUUID(counter, tagger)
+		if err != nil {
+			return nil, err
+		}
+
+		// Get the sealed ID from the IO Provider. If error, all the Identifiers have been found, and the
+		// function can return them.
+		sealedID, err := i.getSealedIdentifier(labelUUID)
+		if err != nil {
+			break
+		}
+
+		decryptedID, err := sealedID.Unseal(labelUUID, cryptor)
+		if err != nil {
+			return nil, err
+		}
+
+		decryptedIDs = append(decryptedIDs, decryptedID.Identifier)
+
+		counter = decryptedID.NextCounter
+	}
+
+	return decryptedIDs, nil
+}
+
+// Delete deletes a keyword/identifier pair from the secure index.
+func (i *SecureIndex) Delete(token, keyword, identifier string) error {
+	identity, err := i.idProvider.GetIdentity(token)
+	if err != nil {
+		return ErrNotAuthenticated
+	}
+	if !identity.Scopes.Contains(id.ScopeIndex) {
+		return ErrNotAuthorized
+	}
+
+	tagger, cryptor, err := i.getTaggerAndCryptor(keyword)
+	if err != nil {
+		return err
+	}
+
+	// A map that keeps track of which labels should be updated after deletions.
+	// It is necessary to first update labels after deletions (as the same keyword/
+	// identifier pair can occur more than once in the secure index).
+	labelUpdates := make(map[uint64]uuid.UUID)
```

Comment:
I'm not sure this is a good idea. As long as we don't have transactions we very easily get into an inconsistent state if we encounter an error along the way. 

I'd suggest the following: instead of deleting say `(f(1),2)`, just call update with `(f(1),3)` and then delete `(f(2),3)`. Even if the delete fails, we still have a consistent state.